### ISSUE TITLE
rpmbuild 4.17 fails with SIGPIPE on FreeBSD 12.3

### ISSUE THREAD

**Issue by wichtounet** [rpm-software-management/rpm#2129]
Hi, 

We just upgraded to RPM 4.17 on FreeBSD 12.3 and we are getting issues with SIGPIPE. 
When run through gdb, we see

```
Thread 1 received signal SIGPIPE, Broken pipe.
Sent by kernel. 
```

This happens in `_write `(call by `__thr_write`). 

We did not have this issue with 4.16. 

Is this a known issue? Is there a workaround?

**Comment by dmnks** [comment]
Ugh, this is strange, and not a known issue so far. 4.17.1 is quite fresh, though...

The closest we went to signal handling in 4.17.1 was perhaps commit 25c4c61a9599590ce4d610621ff0bffde7ae5e82, however it's hard to say if that could cause a regression like this.

Can you reproduce this reliably? If so, mind sharing the steps? Would you be able to git-bisect this with a local RPM checkout? Thanks!

**Comment by wichtounet** [comment]
We are actually using 4.17.0, not 4.17.1. 4.17.1 is not yet available for us. 

We can reproduce it reliably with most of our spec files, but sometimes it fails, and sometimes it does not. But even with 10 retries, we can't build all our RPMs. I am afraid I can't share the spec files.  But I can share the rpmbuild command if that helps a little. 

I can try 4.17.1 with a local build or bisecting from 4.16 to 4.17. Which makes more sense?

**Comment by dmnks** [comment]
Oh, apologies - of course, you mentioned 4.17, not 4.17.1, I misread. In that case, yeah, there's been a few more changes between 4.16 and 4.17 as those are two different major versions.

Bisecting between 4.16 and 4.17 would then make more sense. And do share the rpmbuild command at least, just in case something rings a bell!

**Comment by dmnks** [comment]
> Bisecting between 4.16 and 4.17 would then make more sense. And do share the rpmbuild command at least, just in case something rings a bell!

Practically, this would be (roughly) the log between commits 6e51ca5959d9e5159fdd75c0462c900d342f2cce and 2a03b8fa12319d4650ee3cd1c48d1cd3ab8afe56.

**Comment by wichtounet** [comment]
Do you know which packages I need for lua to build rpm? I have installed `lua52`, `lua53 `and `luajit` and `autogen.sh `still tells me there is no LUA.

**Comment by dmnks** [comment]
Do you have the Lua *development* libraries installed, too? E.g. in Fedora, this is shipped in the `lua-devel` package.

**Comment by dmnks** [comment]
We have a [requirement](https://github.com/rpm-software-management/rpm/blob/f5118bba73c0fb242123ce62c7278d0d3a8cd7e5/configure.ac#L698) for Lua 5.2 at a minimum, but that doesn't seem to be the culprit in your case. I'm not familiar with FreeBSD packaging to suggest specific packages, I'm afraid...

**Comment by DemiMarie** [comment]
Is there a reason that RPM (the command-line tool) does not just install an empty signal handler for SIGPIPE?

**Comment by dmnks** [comment]
We do ignore SIGPIPE during a transaction, but we revert to the default handler during scriplet execution.
@wichtounet, do your spec files have anything in common, such as a scriptlet or a certain common pattern used in them?

**Comment by pmatilai** [comment]
FWIW, there's a known signal related regression in 4.17, fixed in 4.17.1 (see commit 491537593e9fcab2d25b1816fc6bba846c599337). So I'd start with testing 4.17.1 and if that doesn't fix the issue, bisect.

The issue with Lua is that rpm expects Lua pkg-config entry to be called "lua" and not eg "lua52", so you'll need to tweak that in configure.ac. IIRC there's even a ticket on this somewhere, the same issue exists on Debian.

**Comment by wichtounet** [comment]
Our spec files do have scriptlets indeed. I don't know about a common pattern, though. 

Currently, it looks like 4.17.1 is unavailable in our repos, so we will revert to 4.16. I don't have access to a machine that could build this properly.

**Comment by pmatilai** [comment]
I don't know how FreeBSD build system works, but whatever the "recipe" that build 4.17.0 in your repos should work for 4.17.1 by just bumping the number.

**Comment by pmatilai** [comment]
Anyway, we have a pretty good reason to think this is a known regression, fixed in 4.17.1. If it turns out not to be the case, please reopen.
